### ISSUE TITLE
Error using tio.Queue with different num_workers in Windows

### ISSUE THREAD

**Issue by estebvac** [fepegar/torchio#422]
**🐛Bug**

Hi!, thanks for the excellent library. I am trying to integrate it into one of my projects. I am trying to use Queue with different num_workers in windows, but it produces the following error. I can run it using num_workers = 0. 

I am using this in a PC equipped with 32RAM, i7-8700, GPU RTX 2080 and Windows 10. 
This error only occurs in Windows; when I launch the same code in colab it works correctly. Is there any workaround for running it in windows?

**To reproduce**

```python
import torchio as tio
from torch.utils.data import DataLoader
patch_size = 96
queue_length = 300
samples_per_volume = 10
sampler = tio.data.UniformSampler(patch_size)
subject = tio.datasets.Colin27()
subjects_dataset = tio.SubjectsDataset(10 * [subject])
patches_queue = tio.Queue(
    subjects_dataset,
    queue_length,
    samples_per_volume,
    sampler,
    num_workers=4,)
```

```python-traceback
Traceback (most recent call last):
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'Queue.get_subjects_iterable.<locals>.<lambda>'
python-BaseException
Traceback (most recent call last):
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\spawn.py", line 125, in _main
    prepare(preparation_data)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\spawn.py", line 236, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\spawn.py", line 287, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
  File "C:\Users\fzj\anaconda3\lib\runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "C:\Users\fzj\anaconda3\lib\runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "C:\Users\fzj\anaconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\fzj\Google Drive\PHD project\3_src\DL\sli_net\main_loader.py", line 10, in <module>
    patches_queue = tio.Queue(
  File "C:\Users\fzj\anaconda3\lib\site-packages\torchio\data\queue.py", line 146, in __init__
    self.subjects_iterable = self.get_subjects_iterable()
  File "C:\Users\fzj\anaconda3\lib\site-packages\torchio\data\queue.py", line 240, in get_subjects_iterable
    return iter(subjects_loader)
  File "C:\Users\fzj\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 352, in __iter__
    return self._get_iterator()
  File "C:\Users\fzj\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 294, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "C:\Users\fzj\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 801, in __init__
    w.start()
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\context.py", line 326, in _Popen
    return Popen(process_obj)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\popen_spawn_win32.py", line 45, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\spawn.py", line 154, in get_preparation_data
    _check_not_importing_main()
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\spawn.py", line 134, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError: 
  An attempt has been made to start a new process before the
  current process has finished its bootstrapping phase.

  This probably means that you are not using fork to start your
  child processes and you have forgotten to use the proper idiom
  in the main module:

  if __name__ == '__main__':
      freeze_support()
      ...

  The "freeze_support()" line can be omitted if the program
  is not going to be frozen to produce an executable.
python-BaseException

Process finished with exit code -1
```

**Comment by fepegar** [comment]
Hi, @estebvac. Thanks for your kind words.

1. Have you tried using the syntax suggested in the error message? See [this SO answer](https://stackoverflow.com/a/18205006/3956024)
2. What version of SimpleITK are you using? You can check that running `conda list`

**Comment by fepegar** [comment]
Also, can you try this code?

```python
import torch
loader = torch.utils.data.DataLoader(list(range(10)), num_workers=2)
iter(loader).next()
```

**Comment by estebvac** [comment]
Thanks @fepegar for your quick answer, I have simpleitk 1.2.4 installed, I tried your code as in this script:

```
  import torch
  import torchio as tio
  def main():
      import torch
      loader = torch.utils.data.DataLoader(list(range(10)), num_workers=2)
      iter(loader).next()
  
      patch_size = 96
      queue_length = 300
      samples_per_volume = 10
      sampler = tio.data.UniformSampler(patch_size)
      subject = tio.datasets.Colin27()
      subjects_dataset = tio.SubjectsDataset(10 * [subject])
      patches_queue = tio.Queue(
          subjects_dataset,
          queue_length,
          samples_per_volume,
          sampler,
          num_workers=4, )
  
  if __name__ == '__main__':
      main()
```
  


PyTorch with DataLoader worked correctly. But Queue still does not. 

Traceback (most recent call last):
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'Queue.get_subjects_iterable.<locals>.<lambda>'
python-BaseException
Traceback (most recent call last):
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\spawn.py", line 126, in _main
    self = reduction.pickle.load(from_parent)
EOFError: Ran out of input
python-BaseException


Could this problem be related to this [thread](https://discuss.pytorch.org/t/cant-pickle-local-object-dataloader-init-locals-lambda/31857) it seems that Pytorch in windows does not support lambda functions, could this be related to this part of your code?:
```
subjects_loader = DataLoader(
            self.subjects_dataset,
            num_workers=self.num_workers,
            collate_fn=lambda x: x[0],
            shuffle=self.shuffle_subjects,
        )
```

**Comment by fepegar** [comment]
I guess this is related to #366.

In that code you just shared, you're not using the queue, you only instantiate it. Did the error happen after running that code only? Is that the full error message?

I don't have access to Windows at the moment. I feel like I've tried before, but could you please try defining that function before, instead of using a lambda? I.e. `def f(x): return x[0]` and then `collate_fn=f`.

Lastly, please use syntax highlighting for your code blocks. I have edited your comments adding `python` and `python-traceback` where necessary, as an example.

**Comment by estebvac** [comment]
The error occurs when I instantiate Queue. The full error is as follows:
``` python-traceback
Traceback (most recent call last):
  File "C:/Users/fzj/Google Drive/PHD project/3_src/DL/sli_net/main_loader.py", line 24, in <module>
    main()
  File "C:/Users/fzj/Google Drive/PHD project/3_src/DL/sli_net/main_loader.py", line 16, in main
    patches_queue = tio.Queue(
  File "C:\Users\fzj\anaconda3\lib\site-packages\torchio\data\queue.py", line 146, in __init__
    self.subjects_iterable = self.get_subjects_iterable()
  File "C:\Users\fzj\anaconda3\lib\site-packages\torchio\data\queue.py", line 240, in get_subjects_iterable
    return iter(subjects_loader)
  File "C:\Users\fzj\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 352, in __iter__
    return self._get_iterator()
  File "C:\Users\fzj\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 294, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "C:\Users\fzj\anaconda3\lib\site-packages\torch\utils\data\dataloader.py", line 801, in __init__
    w.start()
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\context.py", line 326, in _Popen
    return Popen(process_obj)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\popen_spawn_win32.py", line 93, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\fzj\anaconda3\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'Queue.get_subjects_iterable.<locals>.<lambda>'

Process finished with exit code 1
```

**Comment by fepegar** [comment]
Thanks that error message helped me understand a bit more.

Would you be able to try what I asked above (replacing the `lambda` with a defined function)?

**Comment by estebvac** [comment]
Yep, I managed to make it work, changing the part in your code that contains `lambda`, in the code of [queue.py](https://github.com/fepegar/torchio/blob/master/torchio/data/queue.py)

```python

    def get_subjects_iterable(self) -> Iterator:
        # I need a DataLoader to handle parallelism
        # But this loader is always expected to yield single subject samples
        self._print(
            '\nCreating subjects loader with', self.num_workers, 'workers')
        subjects_loader = DataLoader(
            self.subjects_dataset,
            num_workers=self.num_workers,
            collate_fn=self.lambdax_0,
            shuffle=self.shuffle_subjects,
        )
        return iter(subjects_loader)

    @staticmethod
    def lambdax_0(x):
        return x[0]
```
With this change this [code](https://github.com/fepegar/torchio/issues/422#issuecomment-765555344) worked!

**Comment by fepegar** [comment]
Many thanks for trying that. Hopefully, this will be solved after merging #423.

**Comment by fepegar** [comment]
Hopefully, the changes in #423, which are available on `v0.18.21`, will have fixed this issue.

@all-contributors please add @estebvac for bug.
